### PR TITLE
SiteSettingsSeoFrom: refactor away from `UNSAFE_cWRP`

### PR DESCRIPTION
**Pull request by flootr**
#### Changes proposed in this Pull Request

* `SiteSettingsFormSEO`: refactor away from `UNSAFE_cWRP`

#### Testing instructions

* Open `/marketing/traffic/:site` for a site with a business plan
* Make sure _Enable SEO Tools_ is checked
* For _Page Title Structure_ and _Website Meta_ change settings and save them
* Make sure they're persisted and everything behaves as on prod
* Also try the error case by blocking network requests to `/sites/:siteId/settings`

Related to #58453

### CODE REVIEW

File: client/my-sites/site-settings/seo-settings/form.jsx [hunk 1]
```diff
@@ -62,91 +65,59 @@ function getGeneralTabUrl( slug ) {
...
 
-		// Don't update state for fields the user has edited
-		nextState = omit( nextState, Array.from( dirtyFields ) );
+	handleSubmitSuccess() {
+		this.props.markSaved();
+		this.props.requestSiteSettings( this.props.siteId );
+		this.refreshCustomTitles();
 
-		this.setState( nextState );
+		if ( this._mounted ) {
```
Comment (tyxla) [inline client/my-sites/site-settings/seo-settings/form.jsx:112]: Instead of relying on this legacy approach, can we use a ref instead? A ref's `.current` will be set to `null` when the component unmounts, so no need to manage an extra `_mounted` property manually.

File: client/my-sites/site-settings/seo-settings/form.jsx [hunk 2]
```diff
@@ -214,7 +181,11 @@ export class SeoForm extends Component {
 			( format ) => ( Array.isArray( format ) && 0 === format.length ? '' : format )
 		);
 
-		this.props.saveSiteSettings( siteId, updatedOptions );
+		this.props.saveSiteSettings( siteId, updatedOptions ).then( ( res ) => {
+			if ( res.updated ) {
```
Comment (tyxla) [inline client/my-sites/site-settings/seo-settings/form.jsx:179]: 🤔  Why was this addition necessary?

File: client/my-sites/site-settings/seo-settings/form.jsx [hunk 3]
```diff
@@ -512,4 +485,14 @@ const mapDispatchToProps = {
 		recordTracksEvent( 'calypso_seo_tools_front_page_meta_updated', properties ),
 };
 
-export default connect( mapStateToProps, mapDispatchToProps )( protectForm( localize( SeoForm ) ) );
+const withCurrentSiteKey = createHigherOrderComponent(
```
Comment (tyxla) [inline client/my-sites/site-settings/seo-settings/form.jsx]: Instead of having to create a HoC just to key by site ID, can we just pass the site ID [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/marketing/traffic/index.js#L51) as a prop?

File: client/my-sites/site-settings/seo-settings/form.jsx [hunk 4]
```diff
@@ -62,91 +65,59 @@ function getGeneralTabUrl( slug ) {
...
+				props.selectedSite.options?.advanced_seo_front_page_description,
+				state.frontPageMetaDescription
+			)
+		) {
+			nextState.frontPageMetaDescription =
+				props.selectedSite.options?.advanced_seo_front_page_description;
 		}
 
-		// if we are changing sites, everything goes
-		if ( prevSite.ID !== nextSite.ID ) {
```
Comment (tyxla) [inline client/my-sites/site-settings/seo-settings/form.jsx:115]: Good idea to replace this with the `key={ siteId }` approach 👍

File: client/my-sites/site-settings/seo-settings/form.jsx [hunk 5]
```diff
@@ -62,91 +65,59 @@ function getGeneralTabUrl( slug ) {
...
-			this.setState( { isSubmittingForm: false } );
+		if (
+			! dirtyFields.has( 'seoTitleFormats' ) &&
+			! isEqual( props.storedTitleFormats, state.seoTitleFormats )
+		) {
+			nextState.seoTitleFormats = props.storedTitleFormats;
 		}
 
-		// save error
-		if ( this.state.isSubmittingForm && nextProps.saveError ) {
```
Comment (tyxla) [inline client/my-sites/site-settings/seo-settings/form.jsx:104]: What's happening with this error notice? Are we removing it intentionally?

File: client/my-sites/site-settings/seo-settings/form.jsx [hunk 6]
```diff
@@ -62,91 +65,59 @@ function getGeneralTabUrl( slug ) {
...
-			this.setState( { isSubmittingForm: false } );
-			this.props.errorNotice(
-				translate( 'There was a problem saving your changes. Please, try again.' ),
-				{
-					id: 'seo-settings-form-error',
-				}
-			);
+		if (
```
Comment (tyxla) [inline client/my-sites/site-settings/seo-settings/form.jsx:85]: I wonder if we could clean that derived state up if we make the form fields uncontrolled and treat the remote server values as the canonical values? IIRC we do that in other areas of the site settings management code.

**Review by tyxla — COMMENTED** [review]
Thanks for working on this one! I wonder what are your thoughts on whether there are a few more cleanup / improvement opportunities here.

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 5 above)
Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:104]: It used to be dispatched twice. One from the component and one from within the thunk. So I removed this one.

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 2 above)
Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:179]: The `saveSiteSettings` thunk doesn't rethrow in case there's an error and we don't want to call `handleSubmitSuccess` in case there's an error.

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 3 above)
Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx]: I don't know. Wouldn't that mean we don't just get a new `SiteSettingsSEOForm` but a new `AsyncLoad` wrapper component as well? I've gotta try.

Comment (tyxla) [reply client/my-sites/site-settings/seo-settings/form.jsx]: I think we should be good; we shouldn't get a new `AsyncLoad` wrapper component. `AsyncLoad` will load the component only once, and will pass the new props down to it if they get updated at some point.

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 2 above)
Comment (tyxla) [reply client/my-sites/site-settings/seo-settings/form.jsx:179]: Nice 👍

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 5 above)
Comment (tyxla) [reply client/my-sites/site-settings/seo-settings/form.jsx:104]: Gotcha 👍

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 1 above)
Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:112]: Hm, that doesn't seem to be true at least for class components (haven't checked functional ones). I tried to replicate the behavior in this [sandbox](https://codesandbox.io/s/exciting-ramanujan-6yvvr) and if you hit _Hide_ before the async "task" is done you still get an error in the console. I may be misunderstanding though. Let me know what you think.

Comment (tyxla) [reply client/my-sites/site-settings/seo-settings/form.jsx:112]: Yes, that's expected because you haven't added assigned that ref to an element.

If you assign `ref={ this.mounted }` to an element, React will set that ref to `null` when the component unmounts. Let me know if that makes sense.

Note that this behavior is also documented [here](https://reactjs.org/docs/refs-and-the-dom.html#accessing-refs): 

> React will assign the `current` property with the DOM element when the component mounts, and assign it back to `null` when it unmounts. ref updates happen before `componentDidMount` or `componentDidUpdate` lifecycle methods.

Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:112]: Yes, totally makes sense 🤦🏻 – although it ~~still~~ also feels like a workaround to me, happy to change it though.

Comment (tyxla) [reply client/my-sites/site-settings/seo-settings/form.jsx:112]: IMHO it is as much workaround as the `key={ siteId }` that we often do to reset the component when the site is switched. Not having strong feelings of course.

Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:112]: Fixed via https://github.com/Automattic/wp-calypso/pull/59940/commits/0f1665cfcfbe4779f460fef1ff0e2e22587d913a

Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:112]: > IMHO it is as much workaround as the `key={ siteId }` that we often do to reset the component when the site is switched. Not having strong feelings of course.

totally agree 👍🏻

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 3 above)
Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx]: Fixed via https://github.com/Automattic/wp-calypso/pull/59940/commits/cc01884b2e95514ec6aafb123f8bc7e14f57d9fc – I don't think `key` is passed down but it's still working as expected and removes the need for an HOC.

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 6 above)
Comment (flootr) [reply client/my-sites/site-settings/seo-settings/form.jsx:85]: I'd rather explore that in a follow up to get this one in if you're okay with it.

**Comment by flootr** [comment]
@tyxla I think this one is ready for another review.

File: client/my-sites/site-settings/seo-settings/form.jsx (hunk 6 above)
Comment (tyxla) [reply client/my-sites/site-settings/seo-settings/form.jsx:85]: Sure, fine to explore in another PR, and consider that it's totally optional as well.

**Review by tyxla — APPROVED** [review]
LGTM 👍
